### ISSUE TITLE
Question/RFC: encode the entity type in its URN

### ISSUE THREAD

**Issue by massimiliano-mantione** [radicle-dev/radicle-link#171]
A foreword: note that in this reasoning "entity" means anything that is handled by the `Entity` type in the code. Other "things" like issues, issue comments, and so on are not considered here.

While writing the entity resolver I finally stumbled on the problem of entity discoverability, and when you discover entities how to distinguish between `User` from `Project` instances.

There are two main ways to tackle this problem:

1. Have an enum that can be either a `User` or a `Project` and employ some serde trickery so that any user or project blob can also be deserialized into the union type if needed. This is not that hard, and is what I had in mind so far.

2. Encode the entity "kind" into the "path" section of the URN, so that when we discover we can filter on it using the git glob API for refs, and then we can deserialize knowing the correct type.

Both approaches would get the job done but their implementation is totally different.

The second one would likely lead to cleaner code but we would need to decide the convention for paths pointing to entities of different kinds.

Right now the canonical path is `rad/id`.
The two obvious alternatives would be `rad/user` and `rad/project` (or `rad/usr` and `rad/prj`, and then `rad/org` when we'll have organizations as entities).

But, if we go this way, we should also spend a moment thinking about another requirement we had pending, of being able to store more than one entity in a single repository.
IIRC the idea was to be able to create a project and manage the identities of its maintainers directly inside the same repo.

We should decide if this is really a requirement or not because it complicates entity discovery significantly.
Right now an entity has an ID which is a hash, and listing all the known entities in the monorepo means listing the monorepo namespaces, which is quite simple with a glob.

If each namespace could contain more entities we'd have a dilemma:

* Would these sub-entities have a hash ID of their own?
  - if yes, why cannot we find them as namespaces at the global level?
  - if no, their identity is definitely scoped into this repo, and their usefullness as entities is questionable (the whole TUF update dance does not necessarily make sense inside a single repo...)
* How do we discover them so we can list "all known users" in the UI?

One way to solve the discoverability problem would be to have sub-namespaces inside the rad namespace, like:

```
/rad
    /usr   <--- if present this is "self", the canonical entity of this repo, which is a user
        /foo   <--- another user hosted here
        /bar   <--- another user hosted here
    /prj   <--- if present this is "self", the canonical entity of this repo, which is a project
        /foo   <--- another project hosted here
        /bar   <--- another project hosted here
```

This approach is extensible and would likely work.
It would make it mandatory to have a "path" in the URN, unless we decide on a sensible default, which could be `rad/prj` (making the path mandatory only for entities that are not projects).

What do you think?

**Comment by kim** [comment]
There is a third approach, which is preferable imho: add a discriminator to the URN components, like:

    rad:<protocol>:<type>:<hash>[<path>]

e.g.

    rad:git:project:deadbeef

If you re-read the [monorepo RFC](docs/rfc/identity_resolution.md), you will notice that it accounts for _certifier identities_:

* The identity of the git repository is **always** stored at `rad/id`. There is one, and only one, identity of a repo.
* Additional identities needed for verification (ie. certifiers) are stored at `rad/ids/*`. Those are named after their hashes, and are _symbolic refs_ to the `rad/id` branches of their respective top-level namespace (for the owned branch hierarchy, not remotes). There is a guarantee that the `link` replication protocol will implicitly create top-level namespaces, but an exported git repository can still be verified by traversing `rad/ids/*`.

Please correct me if I'm wrong, but I believe this covers all uses you mentioned. 

The problem of distinguishing entity types from an iterator over namespaces (or just URNs, for that matter) _without attempting to parse the entity data_ is orthogonal. I said before that I would prefer to add the entity type to the URN, because this renders it "machine readable" (in the semantic web sense). Interpreting the path component according to some conventions seems much more brittle than using a `:`-delimited component.

**Comment by FintanH** [comment]
I'm also for storing the `type` of the entity in the `RadUrn`. 

One thing that I don't see covered above is how to list all the projects/users in the monorepo. Which I think can be achieved by nesting the `type` in the namespace. That way we can fetch, for example projects, by using the glob: `namespaces/**/refs/namespaces/user`.

**Comment by massimiliano-mantione** [comment]
> The problem of distinguishing entity types from an iterator over namespaces (or just URNs, for that matter) _without attempting to parse the entity data_ is orthogonal.

This is probably the most important bit, but note that iterating over namespaces vs URNs are quite different things.

Even if we added a "type discriminator" at the beginning of URNs, the monorepo does not store URNs, only refs inside namespaces.

The actual problem is be _filtering_ entities of a given type while listing them.
The monorepo is not storing the types, so without a cache the only "listing" we can do is iterating over all the namespaces inspecting all the blobs and filtering the ones we do not want.
Having the type inside the URN does not help here.
Only having it in the path section of the URN would help, and this is the question I was asking here.

You say that the two concerns are orthogonal.

However, even if if we had the "type discriminator" in the URN (not in the path section), we should implement the "deserialize into union type" piece of code anyway (to solve the "listing" problem).
Once you already have code that can parse an entity and return a union type, checking the entity type is error handling as usual which we cannot skip anyway (the blob could disagree with the URN).

This is why I was proposing, **if** we want to add a "type discriminator" to URNs, to do it in the path: because then we could use globs to filter entities by type.
I know that in the past I had a different opinion but now that I am facing the code I see things differently :-)

Finally, about storing more entities inside a single repo, I had misunderstood the requirement.
Now I see what the `refs/ids` part is doing and I understand that it does not impact entity discovery so it's not relevant here.

Summing up, I would implement the "deserialize to union type" part and base URN discovery and resolution on that.

Does it make sense?

**Comment by kim** [comment]
> Even if we added a "type discriminator" at the beginning of URNs, the monorepo does not store URNs, only refs inside namespaces.

I see, yeah, but we can change that. 

The problem I have with encoding it in the URN path is that we fixed its semantics to denote "ref relative to identity". Thus, `deadbeef/rad/xyz` means a different thing than `deadbeef/rad/abc`, _even though_ `deadbeef` is (supposedly) globally unique and corresponds to one-and-only-one identity document.

The simplest thing is probably to change the namespace encoding, but that has the same problem: `deadbeef.project` and `deadbeef.user` are two different things, when only one should exist.

Another option is to turn `rad/id` into a symref pointing to a ref which can be globbed (eg. `rad/usr`, `rad/prj`). If we only have a `rad/id` `Reference` handy, we can peek at its `symbolic_target`. For some reason, however, `upload-pack` currently only reports the `HEAD` symref as a capability (cf. https://github.com/git/git/commit/7171d8c15f919c760d52f814a0aee1b1253385b1), so we still need to know (or guess) what the entity type shall be when cloning. That's were a value-level discriminator in the URN would be helpful.

Yet another trick would be to keep the type information outside the `refs/` hierarchy (possibly closest to what you mean by "cache"): just like `HEAD`, `FETCH_HEAD`, etc., they can live in `$GIT_DIR` directly. So, for example, we could create `$GIT_DIR/rad/usr/<hash> -> refs/namespaces/<hash>/rad/id` after creating or cloning. The type of an entity can never change, so this is a one-off action. We need to check if `git2::Repository::references_glob` can deal with that, but if not it's a bug.

> we should implement the "deserialize into union type" piece of code anyway

I don't exactly understand what you mean here: we don't need this if we know what we'd expect, or no?

**Comment by massimiliano-mantione** [comment]
In turns out that we do not know what to expect.

There main problematic use case is listing entities in the monorepo.
In this case, since in the monorepo we do not store URNs, even if we did put the entity type in the URN we would be out of luck: the monorepo does not state what each entity is.

The low level "entity listing" API implemented in #170 has exactly this problem: it cannot distinguish different entity types before attempting deserialization.

The idea is that persisting the entity type in the repo, to be significantly useful, should allow filtering between different entities by globbing their git refs.
But to do this the information would need to be encoded in the "path" portion of the URN and not at the beginning.
Mandating different paths for different entities would have its share of downsides, so for now we decided to go for the simplest approach which is having this union type (implemented as `GenericEntity` in #170).

**Comment by FintanH** [comment]
> since in the monorepo we do not store URNs

I don't think this true. A `RadUrn` maps to a reference in the monorepo.
`rad:git:hwd1yredksthny1hht3bkhtkxakuzfnjxd8dyk364prfkjxe4xpxsww3try/rad/id` maps to `refs/namespaces/hwd1yredksthny1hht3bkhtkxakuzfnjxd8dyk364prfkjxe4xpxsww3try/refs/rad/id`

So I don't think there's anything stopping us from saying what the entity type is in the RadUrn _or_ the monorepo. We would just have to decide how it maps to a ref.

The listing would still be a problem, in the case of wanting to list multiple types of entities. So, in that case, we would need an `enum`, yes.

> But to do this the information would need to be encoded in the "path" portion of the URN and not at the beginning.

I could see it being a nested namespace, e.g.:
```
refs/namespaces/hwd1yredksthny1hht3bkhtkxakuzfnjxd8dyk364prfkjxe4xpxsww3try/refs/namespaces/user/refs/rad/id
```
But maybe this would cause complications? I'm not sure off the top of my head.

**Comment by massimiliano-mantione** [comment]
> But maybe this would cause complications? I'm not sure off the top of my head.

This is exactly the case: it would force semantics on the "path" section of the URN.

The original proposal to add type info to URNs wanted to put them at the "head" (near the `rad:` part) to avoid this kind of issue.
But that part of the URN is not reflected in the git repo.

So, this issue was opened asking "should we encode the entity type in the URN **in the path section**", and the conclusion was "no, it's better not".

**Comment by FintanH** [comment]
> This is exactly the case: it would force semantics on the "path" section of the URN.

That's not what I'm saying. The `namespaces/user` is not part of the `path` of the `RadUrn`, just in the same way the `id` hash isn't.

Basically my proposal is whether we'd be interested in:
```rust
pub struct RadUrn {
    pub id: Hash,
    pub proto: Protocol,
    pub entity_type: Option<EntityType>,
    pub path: Path,
}

impl Display for RadUrn {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "rad:{}:{}", self.proto.nss(), self.id)?;
        if let Some(entity_type) {
            write!(f, "namespaces/{}", entity_type)?;
        }

        if !self.path.is_empty() {
            write!(
                f,
                "/{}",
                percent_encode(self.path.as_bytes(), PATH_PERCENT_ENCODE_SET)
            )?;
        }

        Ok(())
    }
}


```
